File: samples/graph-pinned-messages/python/controller/pin_message_controller.py
#!/usr/bin/env python3
# Copyright (c) Microsoft Corporation. All rights reserved.
# Licensed under the MIT License.

from flask import Blueprint, request, jsonify
from utils.graph_client import GraphClient
from utils.token_exchange import exchange_sso_token_for_graph_token

chat_api = Blueprint('chat_api', __name__)
delegated_token = ""

# Retrieves pinned and recent messages from a Teams chat using Graph API.
@chat_api.route('/getGraphAccessToken', methods=['GET'])
def get_graph_access_token():
    global delegated_token
    sso_token = request.args.get('ssoToken')
    chat_id = request.args.get('chatId')

    if not sso_token or not chat_id:
        return jsonify({'error': 'Missing ssoToken or chatId'}), 400

    try:
        access_token = exchange_sso_token_for_graph_token(sso_token)
        if not access_token:
            return jsonify({'error': 'Failed to exchange token'}), 401
        delegated_token = access_token
        graph_client = GraphClient(access_token)
        try:
            pinned_messages = graph_client.get_pinned_messages(chat_id)
        except Exception as e:
            pinned_messages = {"value": []}
        recent_messages = graph_client.get_recent_messages(chat_id)
        message_list = []
        for message in recent_messages.get("value", []):
            message_id = message.get('id', 'NO_ID')
            message_type = message.get('messageType', 'unknown')
            if message_type in ["message", "unknownFutureValue"] or message_type is None:
                body = message.get("body", {})
                content = body.get("content", "") if body else ""
                if content:
                    import re
                    clean_content = re.sub(r'<[^>]+>', '', content)
                    if clean_content.strip():
                        message_details = {
                            "id": message_id,
                            "value": clean_content.strip()
                        }
                        message_list.append(message_details)

        pinned_values = pinned_messages.get('value', [])
        if pinned_values and len(pinned_values) > 0:
            pinned_msg = pinned_values[0]
            pinned_content = pinned_msg.get('message', {}).get('body', {}).get('content', '')
            import re
            clean_pinned_content = re.sub(r'<[^>]+>', '', pinned_content)
            response_data = {
                "id": pinned_msg.get('id', ''),
                "message": clean_pinned_content,
                "messages": message_list
            }
        else:
            response_data = {
                "id": '',
                "message": '',
                "messages": message_list
            }

        return jsonify(response_data)
    except Exception as e:
        return jsonify({'error': str(e)}), 500

# Pins a specific message in a Teams chat.
@chat_api.route('/pinMessage', methods=['GET'])
def pin_new_message():
    sso_token = request.args.get('ssoToken')
    chat_id = request.args.get('chatId')
    message_id = request.args.get('messageId')

    if not sso_token:
        return jsonify({'error': 'Missing ssoToken'}), 400
    if not chat_id:
        return jsonify({'error': 'Missing chatId'}), 400
    if not message_id:
        return jsonify({'error': 'Missing messageId. Please select a message to pin.'}), 400

    try:
        access_token = exchange_sso_token_for_graph_token(sso_token)
        if not access_token:
            return jsonify({'error': 'Failed to exchange token'}), 401
        graph_client = GraphClient(access_token)
        result = graph_client.pin_message(chat_id, message_id)
        return '', 204
    except Exception as e:
        return jsonify({'error': str(e)}), 500

# Unpins a message from a Teams chat.
@chat_api.route('/unpinMessage', methods=['GET'])
def unpin_message():
    sso_token = request.args.get('ssoToken')
    chat_id = request.args.get('chatId')
    pinned_message_id = request.args.get('pinnedMessageId')

    if not sso_token:
        return jsonify({'error': 'Missing ssoToken'}), 400

    try:
        access_token = exchange_sso_token_for_graph_token(sso_token)
        if not access_token:
            return jsonify({'error': 'Failed to exchange token'}), 401
        graph_client = GraphClient(access_token)
        result = graph_client.unpin_message(chat_id, pinned_message_id)
        return '', 204
    except Exception as e:
        return jsonify({'error': str(e)}), 500

File: samples/TeamsJS/graph-pinned-messages/nodejs/controller/pin-message-controller.js
// Copyright (c) Microsoft Corporation. All rights reserved.
// Licensed under the MIT License.

const { SimpleGraphClient } = require('../simpleGraphClient');
const { polyfills } = require('isomorphic-fetch');
const msal = require('@azure/msal-node');

var delegatedToken = "";

const getGraphAccessToken = async (req, res) => {
    var tid = process.env.MicrosoftAppTenantId;
    var token = req.query.ssoToken;
    var chatId = req.query.chatId;
    var scopes = ["https://graph.microsoft.com/User.Read"];

  // Creating MSAL client
  const msalClient = new msal.ConfidentialClientApplication({
    auth: {
      clientId: process.env.MicrosoftAppId,
      clientSecret: process.env.MicrosoftAppPassword
    }
  });

  var oboPromise = new Promise((resolve, reject) => {
    msalClient.acquireTokenOnBehalfOf({
      authority: `https://login.microsoftonline.com/${tid}`,
      oboAssertion: token,
      scopes: scopes,
      skipCache: true
    }).then(result => {
      delegatedToken = result.accessToken
      resolve();
    }).catch(error => {
      reject({ "error": error.errorCode });
    });
  });

  oboPromise.then(async function (result) {
    const client = new SimpleGraphClient(delegatedToken);
    try 
    {
    var pinnedMessageResponse = await client.getPinnedMessageList(chatId);
    var recentMessages = await client.getRecentMessageList(chatId);

    var messageList = new Array();
    recentMessages.value.map(message => {
        if(message.messageType == "message") {
            var messageDetails = {
                id: message.id,
                value: message.body.content.replace(/<[^>]+>/g, '')
            }
    
            messageList.push(messageDetails);
        }
    })

    var responseMessageData = {
        id: pinnedMessageResponse.value[0].id,
        message: pinnedMessageResponse.value[0].message.body.content.replace(/<[^>]+>/g, ''),
        messages: messageList
    }

    res.send(responseMessageData);
    }
    catch (ex) 
    {
    console.error(ex);
    res.status(500).send();
    }
    
  }, function (err) {
    console.log(err); // Error: "It broke"
    res.json(err);
  });
}

/** Pin new message in chat. */
const pinNewMessage = async(req, res) => {
    var chatId = req.query.chatId;
    var newMessageId = req.query.messageId;

    try 
    {
       const client = new SimpleGraphClient(delegatedToken);
       var newPinnedMessageResponse = await client.pinNewMessage(chatId, newMessageId);
        res.status(204).send();
    }
    catch (ex) 
    {
        console.error(ex);
        res.status(500).send();
    }
}

/** Unpin existing pinned message. */
const unpinMessage = async(req, res) => {
    var chatId = req.query.chatId;
    var pinnedMessageId = req.query.pinnedMessageId;

    try 
    {
       const client = new SimpleGraphClient(delegatedToken);
       var unpinnedMessageResponse = await client.unpinMessage(chatId, pinnedMessageId);
        res.status(204).send();
    }
    catch (ex) 
    {
        console.error(ex);
        res.status(500).send();
    }
}

module.exports = {
    getGraphAccessToken,
    pinNewMessage,
    unpinMessage
};